src/goto-instrument/contracts/contracts.cpp

```diff
@@ -311,137 +358,154 @@ void code_contractst::check_apply_loop_contracts(
 
   // Insert instrumentation
   // This must be done before havocing the write set.
-  // ^^^ FIXME this is not true ^^^
-  insert_before_swap_and_advance(
-    goto_function.body, loop_head, snapshot_instructions);
-
-  optionalt<cfg_infot> cfg_empty_info;
+  // FIXME: this is not true for write set targets that
+  // might depend on other write set targets.
```

Comment:
Is there a way to check if we are in this case with the specified or inferred targets ?

Reply:
The widening part is not yet fixed.

I could add a separate commit to fix this PR, but is this still high priority given that we double check all inferred targets anyway by instrumenting the loop?

Reply:
Right now loop contracts still use the `havoc_assigns_targetst` class because it contains the widening logic.
 
Widening right after the inference would allow to use the new class `havoc_assigns_clause_targetst` which shares the CAR snapshotting logic with assigns clause checking.